<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="K2Taxonomer">
<title>K2Taxonomer Dashboards • K2Taxonomer</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="K2Taxonomer Dashboards">
<meta property="og:description" content="K2Taxonomer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">K2Taxonomer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/01_K2Taxonomer_singlecell.html">Running K2Taxonomer on Single-cell RNA Sequencing Data</a>
    <a class="dropdown-item" href="../articles/02_K2Taxonomer_bulk.html">Running K2Taxonomer on Bulk Expression Data</a>
    <a class="dropdown-item" href="../articles/03_K2Taxonomer_dashboard.html">K2Taxonomer Dashboards</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/montilab/K2Taxonomer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>K2Taxonomer Dashboards</h1>
                        <h4 data-toc-skip class="author">Eric R.
Reed</h4>
                        
            <h4 data-toc-skip class="date">March 25, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/K2Taxonomer/blob/HEAD/vignettes/03_K2Taxonomer_dashboard.Rmd" class="external-link"><code>vignettes/03_K2Taxonomer_dashboard.Rmd</code></a></small>
      <div class="d-none name"><code>03_K2Taxonomer_dashboard.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>To facilitate the comprehensive interrogation the output of <em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em> this
package includes functionality for generating interactive dashboards,
which include full compendia of results <span class="citation">(Reed and
Monti 2021)</span>. In addition to the partition-level molecular
comparisons included in these dashboards, they include functionality for
performing molecular comparisons of gene expression and gene set
enrichment on user-specified sets of two-or-more subgroups. In this
vignette we describe the steps for creating, editing, and navigating
interactive dashoboards from the output of the <em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em>
workflow <span class="citation">(Reed and Monti 2021)</span>. For a more
detailed description for running <em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em>
visit the vignettes describing single-cell and single-cell workflows <a href="https://montilab.github.io/K2Taxonomer/articles/01_K2Taxonomer_singlecell">here</a>,
and <a href="https://montilab.github.io/K2Taxonomer/articles/02_K2Taxonomer_bulk">here</a>,
respectively.</p>
</div>
<div class="section level2">
<h2 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h2>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## K2Taxonomer package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://montilab.github.io/K2Taxonomer/">K2Taxonomer</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Seurat package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For drawing dendrograms</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://andrie.github.io/ggdendro/" class="external-link">ggdendro</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="run-k2taxonomer-workflow-single-cell-example">Run K2Taxonomer workflow (Single-cell Example)<a class="anchor" aria-label="anchor" href="#run-k2taxonomer-workflow-single-cell-example"></a>
</h2>
<div class="section level3">
<h3 id="read-in-single-cell-rnaseq-data">Read in single-cell RNAseq data<a class="anchor" aria-label="anchor" href="#read-in-single-cell-rnaseq-data"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ifnb_small"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="read-in-gene-sets-for-subgroup-annotation">Read in gene sets for subgroup annotation<a class="anchor" aria-label="anchor" href="#read-in-gene-sets-for-subgroup-annotation"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"cellMarker2_genesets"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-necessary-data-objects">Get necessary data objects<a class="anchor" aria-label="anchor" href="#get-necessary-data-objects"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Integrated expression matrix used for clustering data</span></span>
<span><span class="va">integrated_expression_matrix</span> <span class="op">&lt;-</span> <span class="va">ifnb_small</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">integrated</span><span class="op">$</span><span class="va">scale.data</span></span>
<span></span>
<span><span class="co">## Normalized expression matrix to be used for downstream analyses</span></span>
<span><span class="va">normalized_expression_matrix</span> <span class="op">&lt;-</span> <span class="va">ifnb_small</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">SCT</span><span class="op">$</span><span class="va">data</span></span>
<span></span>
<span><span class="co">## Profile-level information</span></span>
<span><span class="va">cell_data</span> <span class="op">&lt;-</span> <span class="va">ifnb_small</span><span class="op">@</span><span class="va">meta.data</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-k2taxonomer-workflow">Run K2Taxonomer workflow<a class="anchor" aria-label="anchor" href="#run-k2taxonomer-workflow"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Initialize `K2` object</span></span>
<span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/K2preproc.html">K2preproc</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">integrated_expression_matrix</span>,</span>
<span>                   eMatDS <span class="op">=</span> <span class="va">normalized_expression_matrix</span>,</span>
<span>                   colData <span class="op">=</span> <span class="va">cell_data</span>,</span>
<span>                   cohorts<span class="op">=</span><span class="st">"cell_type"</span>,</span>
<span>                   nBoots <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                   clustFunc <span class="op">=</span> <span class="st">"cKmeansDownsampleSqrt"</span>,</span>
<span>                   genesets <span class="op">=</span> <span class="va">cellMarker2_genesets</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform recursive partitioning</span></span>
<span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/K2tax.html">K2tax</a></span><span class="op">(</span><span class="va">K2res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Partition-level differential gene expression</span></span>
<span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDGEmods.html">runDGEmods</a></span><span class="op">(</span><span class="va">K2res</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Perform Fisher Exact Test based over-representation analysis</span></span>
<span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runFISHERmods.html">runFISHERmods</a></span><span class="op">(</span><span class="va">K2res</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Perform single-sample gene set scoring</span></span>
<span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScoreGeneSets.html">runScoreGeneSets</a></span><span class="op">(</span><span class="va">K2res</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Perform partition-level differential gene set score analysis</span></span>
<span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDSSEmods.html">runDSSEmods</a></span><span class="op">(</span><span class="va">K2res</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="check-partitioning-results">Check partitioning results<a class="anchor" aria-label="anchor" href="#check-partitioning-results"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get dendrogram from K2Taxonomer</span></span>
<span><span class="va">dendro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/K2dendro.html">K2dendro</a></span><span class="op">(</span><span class="va">K2res</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot dendrogram</span></span>
<span><span class="fu"><a href="https://andrie.github.io/ggdendro/reference/ggdendrogram.html" class="external-link">ggdendrogram</a></span><span class="op">(</span><span class="va">dendro</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="creating-a-dashboard">Creating a dashboard<a class="anchor" aria-label="anchor" href="#creating-a-dashboard"></a>
</h2>
<p>We generate the interactive dashboards with the
<code><a href="../reference/K2dashboard.html">K2dashboard()</a></code> function. This function will create a
directory with two files. These two files include the interactive
<em>RMarkdown</em> (.Rmd) file created by the <em><a href="https://CRAN.R-project.org/package=flexdashboard" class="external-link">flexdashboard</a></em>
package and the R object (.rds) file containing the <code>K2</code>
object created by <em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em>
functions.</p>
<p>Below we shown the <code><a href="../reference/K2dashboard.html">K2dashboard()</a></code> function and each of
its arguments.</p>
<p>The first argument, specifying the <code>K2</code> object is the only
requirement. The remaining three arguments costumize the output, as
follows:</p>
<ul>
<li><p><strong>analysis_name</strong>: Specifies the title to be printed
on the top of the dashboard. Also, specifies the prefix of the name of
the output files. For filenames, spaces are replaced by
underscores.</p></li>
<li><p><strong>about</strong>: Logical specifying whether to include an
about page in the dashboard. If TRUE (default), a third file is written
to the dashboard directory, “<strong>about.md</strong>”. This is an
editable <em>markdown</em> (.md) file, which the user is free to edit,
and includes some default information about where to learn more about
<em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em> and
how to navigate the dashboard. Customizing this file is described in
more detail below.</p></li>
<li><p><strong>output_dir</strong>: This is simply the path to the
directory to write the dashboard directory.</p></li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/K2dashboard.html">K2dashboard</a></span><span class="op">(</span><span class="va">K2res</span>,</span>
<span>            analysis_name<span class="op">=</span><span class="st">"K2Taxonomer Example"</span>,</span>
<span>            about<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>            output_dir<span class="op">=</span><span class="st">"."</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="additional-considerations">Additional considerations<a class="anchor" aria-label="anchor" href="#additional-considerations"></a>
</h3>
<div class="section level4">
<h4 id="the-output-directory">The output directory<a class="anchor" aria-label="anchor" href="#the-output-directory"></a>
</h4>
<p>To prevent the overwriting of dashboard files, in addition to the
<strong>analysis_name</strong>, the name of dashboard directory includes
the date and time. It is formatted as follows:</p>
<center>
<strong>analysis_name</strong>_YEAR_MONTH_DAY_HOUR_MINUTE_SECOND.INTEGER
</center>
<p><strong>Note: Changing the name of this directory does not effect the
dashboard.</strong></p>
</div>
<div class="section level4">
<h4 id="the-about-md-file">The “<strong>about.md</strong>” file<a class="anchor" aria-label="anchor" href="#the-about-md-file"></a>
</h4>
<p><em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em>
allows users to include information about their study in the interactive
dashboards by editing” the “<strong>about.md</strong>” <em>markdown</em>
file. When the “<strong>about.md</strong>” is included in the dashboard
directory, it is read in and the <em>markdown</em> code is compiled
along with the code in the dashboard file.</p>
<p>This file uses <em>markdown</em> syntax specific to the
CRANpkg(“flexdashboard”) “Multiple Pages” layout, which can be reviewed
<a href="https://rmarkdown.rstudio.com/flexdashboard/layouts.html" class="external-link">here</a>.
Mainly, the first two lines of the default “<strong>about.md</strong>”
file.</p>
<pre><code>about
=====================================</code></pre>
<p>Will result in a tab, named <em>About</em>, as the first tab in the
dashboard. Furthermore, lines with headers should start at three hashes,
“###”.</p>
<p>Finally, when including links in the “<strong>about.md</strong>”
file, make sure that clicking on these links opens a new window using
the following:</p>
<pre><code>[TEXT](URL){target="_blank"}</code></pre>
<p>Otherwise, the dashboard will need to reload upon navigating
back.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="navigating-dashboards">Navigating dashboards<a class="anchor" aria-label="anchor" href="#navigating-dashboards"></a>
</h2>
<p><em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em>
dashboards comprise three tabs, described below:</p>
<ul>
<li><p><strong><strong>About</strong></strong>: This optional tab with
which the user can include information about the analysis being
performed. This page is populated by a file,
“<strong>about.md</strong>”, included in the same directory as the
dashboard file. More information about formatting this file can be found
<a href="https://montilab.github.io/K2Taxonomer/articles/K2Taxonomer_dashboard">here</a>.</p></li>
<li><p><strong><strong><em>K2Taxonomer</em> Results</strong></strong>:
This tab includes all of the results generated throughout the <em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em>
workflow, including: <em>partitioning results</em>, <em>partition
stability</em>, <em>gene expression analysis</em>, <em>gene set
enrichment</em>, and <em>phenotypic variable testing (optional)</em>.
More information for how <em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em>
estimates these results can be found <a href="https://montilab.github.io/K2Taxonomer/articles/K2Taxonomer_dashboard">here</a>.</p></li>
<li><p><strong><strong>Compare Multiple</strong></strong>: This tab
allows the user to perform additional molecular comparisons between
subgroups, beyond the partition-level comparisons performed by <em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em>
functions.</p></li>
</ul>
<p>Annotation of each of these tabs is presented below. <strong>Click on
any of these images to view in higher resolution</strong></p>
</div>
<div class="section level2">
<h2 id="about">
<strong><strong>About</strong></strong><a class="anchor" aria-label="anchor" href="#about"></a>
</h2>
<hr>
<p><a href="https://montilab.github.io/K2Taxonomer/articles/page1.jpeg">
<img alt="page1" src="page1.jpeg"></a></p>
<hr>
</div>
<div class="section level2">
<h2 id="k2taxonomer-results">
<strong><strong><em><a href="https://github.com/montilab/K2Taxonomer" class="external-link">K2Taxonomer</a></em>
Results</strong></strong><a class="anchor" aria-label="anchor" href="#k2taxonomer-results"></a>
</h2>
<hr>
<p><a href="https://montilab.github.io/K2Taxonomer/articles/page2.jpeg">
<img alt="page2" src="page2.jpeg"></a></p>
<hr>
<div class="section level3">
<h3 id="subgroup-information">
<strong><em>Subgroup Information</em></strong><a class="anchor" aria-label="anchor" href="#subgroup-information"></a>
</h3>
<hr>
<p><a href="https://montilab.github.io/K2Taxonomer/articles/page3.jpeg">
<img alt="page3" src="page3.jpeg"></a></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="compare-multiple">
<strong><strong>Compare Multiple</strong></strong><a class="anchor" aria-label="anchor" href="#compare-multiple"></a>
</h2>
<hr>
<p><a href="https://montilab.github.io/K2Taxonomer/articles/page4.jpeg">
<img alt="page4" src="page4.jpeg"></a></p>
<hr>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-reed_2020" class="csl-entry">
Reed, Eric R., and Stefano Monti. 2021. <span>“Multi-Resolution
Characterization of Molecular Taxonomies in Bulk and Single-Cell
Transcriptomics Data.”</span> Preprint. <em>Nucleic Acids Research</em>,
November. <a href="https://doi.org/10.1093/nar/gkab552" class="external-link">https://doi.org/10.1093/nar/gkab552</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Eric Reed.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
