<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Replaces gene set results from paired up- and down- gene sets with the difference
of the up-regulated genes and down-regulated genes"><title>Take difference of two paired bi-directional gene set scores — aggregateGeneSetScores • K2Taxonomer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Take difference of two paired bi-directional gene set scores — aggregateGeneSetScores"><meta property="og:description" content="Replaces gene set results from paired up- and down- gene sets with the difference
of the up-regulated genes and down-regulated genes"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">K2Taxonomer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/01_K2Taxonomer_singlecell.html">Running K2Taxonomer on Single-cell RNA Sequencing Data</a>
    <a class="dropdown-item" href="../articles/02_K2Taxonomer_bulk.html">Running K2Taxonomer on Bulk Expression Data</a>
    <a class="dropdown-item" href="../articles/03_K2Taxonomer_dashboard.html">K2Taxonomer Dashboards</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/montilab/K2Taxonomer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Take difference of two paired bi-directional gene set scores</h1>
      <small class="dont-index">Source: <a href="https://github.com/montilab/K2Taxonomer/blob/HEAD/R/aggregateGeneSetScores.R" class="external-link"><code>R/aggregateGeneSetScores.R</code></a></small>
      <div class="d-none name"><code>aggregateGeneSetScores.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Replaces gene set results from paired up- and down- gene sets with the difference
of the up-regulated genes and down-regulated genes</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">aggregateGeneSetScores</span><span class="op">(</span><span class="va">K2res</span>, <span class="va">aggList</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>K2res</dt>
<dd><p>A K2 class object.</p></dd>


<dt>aggList</dt>
<dd><p>A named list where each item is a character vector of length, 
2, comprising the name of the 'up' gene set, and the name of the 'down'
gene set.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>An object of class K2.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Reed ER, Monti S (2021).
“Multi-resolution characterization of molecular taxonomies in bulk and single-cell transcriptomics data.”
<em>Nucleic Acids Research</em>.
<a href="https://doi.org/10.1093/nar/gkab552" class="external-link">doi:10.1093/nar/gkab552</a>
, <a href="https://pubmed.ncbi.nlm.nih.gov/34226941/" class="external-link">https://pubmed.ncbi.nlm.nih.gov/34226941/</a>.

   Hanzelmann S, Castelo R, Guinney J (2013).
“GSVA: gene set variation analysis for microarray and RNA-Seq data.”
<em>BMC Bioinformatics</em>, <b>14</b>(1), 7.
ISSN 1471-2105, <a href="https://doi.org/10.1186/1471-2105-14-7" class="external-link">doi:10.1186/1471-2105-14-7</a>
, <a href="http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-7" class="external-link">http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-7</a>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Read in ExpressionSet object</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biobase" class="external-link">Biobase</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: 'BiocGenerics'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from 'package:stats':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from 'package:base':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, setdiff, table, tapply,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sample.ExpressionSet</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Pre-process and create K2 object</span></span></span>
<span class="r-in"><span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu"><a href="K2preproc.html">K2preproc</a></span><span class="op">(</span><span class="va">sample.ExpressionSet</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No cohorts specified and clustFunc = 'cKmeansDownsampleSqrt' . Setting clustFunc = 'hclustWrapper'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No cohorts specified and recalcDataMatrix = TRUE. Setting recalcDataMatrix = FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No cohorts specified and featMetric = 'F'. Setting featMetric = 'mad'</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run K2 Taxonomer algorithm</span></span></span>
<span class="r-in"><span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu"><a href="K2tax.html">K2tax</a></span><span class="op">(</span><span class="va">K2res</span>,</span></span>
<span class="r-in"><span>            stabThresh<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run differential analysis on each partition</span></span></span>
<span class="r-in"><span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu"><a href="runDGEmods.html">runDGEmods</a></span><span class="op">(</span><span class="va">K2res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Running DGE for partition:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1 / 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2 / 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   3 / 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   4 / 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   5 / 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   6 / 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   7 / 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   8 / 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   9 / 10 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   10 / 10 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Create dummy set of gene sets</span></span></span>
<span class="r-in"><span><span class="va">DGEtable</span> <span class="op">&lt;-</span> <span class="fu"><a href="getDGETable.html">getDGETable</a></span><span class="op">(</span><span class="va">K2res</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">DGEtable</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genesetsMadeUp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    GS1<span class="op">=</span><span class="va">genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>    GS2<span class="op">=</span><span class="va">genes</span><span class="op">[</span><span class="fl">51</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>    GS3<span class="op">=</span><span class="va">genes</span><span class="op">[</span><span class="fl">101</span><span class="op">:</span><span class="fl">150</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run gene set hyperenrichment</span></span></span>
<span class="r-in"><span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu">runGSEmods</span><span class="op">(</span><span class="va">K2res</span>,</span></span>
<span class="r-in"><span>                genesets<span class="op">=</span><span class="va">genesetsMadeUp</span>,</span></span>
<span class="r-in"><span>                qthresh<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in runGSEmods(K2res, genesets = genesetsMadeUp, qthresh = 0.1):</span> could not find function "runGSEmods"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run GSVA on genesets</span></span></span>
<span class="r-in"><span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu">runGSVAmods</span><span class="op">(</span><span class="va">K2res</span>,</span></span>
<span class="r-in"><span>                ssGSEAalg<span class="op">=</span><span class="st">'gsva'</span>,</span></span>
<span class="r-in"><span>                ssGSEAcores<span class="op">=</span><span class="fl">1</span>,</span></span>
<span class="r-in"><span>                verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in runGSVAmods(K2res, ssGSEAalg = "gsva", ssGSEAcores = 1, verbose = FALSE):</span> could not find function "runGSVAmods"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Aggregate paired gene sets</span></span></span>
<span class="r-in"><span><span class="va">aggList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'GS12'</span>, <span class="st">'GS1'</span>, <span class="st">'GS2'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">K2res</span> <span class="op">&lt;-</span> <span class="fu">aggregateGeneSetscores</span><span class="op">(</span><span class="va">K2resaggList</span>, <span class="va">K2res</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in aggregateGeneSetscores(K2resaggList, K2res):</span> could not find function "aggregateGeneSetscores"</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Eric Reed.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

