<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function will generate an object of class, K2.  This will run
pre-processing functions for running K2 Taxonomer procedure."><title>Function to create K2 object for pre-processing — K2preproc • K2Taxonomer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function to create K2 object for pre-processing — K2preproc"><meta property="og:description" content="This function will generate an object of class, K2.  This will run
pre-processing functions for running K2 Taxonomer procedure."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">K2Taxonomer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/01_K2Taxonomer_singlecell.html">Running K2Taxonomer on Single-cell RNA Sequencing Data</a>
    <a class="dropdown-item" href="../articles/02_K2Taxonomer_bulk.html">Running K2Taxonomer on Bulk Expression Data</a>
    <a class="dropdown-item" href="../articles/03_K2Taxonomer_dashboard.html">K2Taxonomer Dashboards</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/montilab/K2Taxonomer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function to create K2 object for pre-processing</h1>
      <small class="dont-index">Source: <a href="https://github.com/montilab/K2Taxonomer/blob/HEAD/R/K2preproc.R" class="external-link"><code>R/K2preproc.R</code></a></small>
      <div class="d-none name"><code>K2preproc.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function will generate an object of class, K2.  This will run
pre-processing functions for running K2 Taxonomer procedure.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">K2preproc</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  cohorts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  eMatDS <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colData <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vehicle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variables <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seuAssay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  seuAssayDS <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  sceAssay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  sceAssayDS <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  block <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  logCounts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Z"</span>, <span class="st">"MEAN"</span><span class="op">)</span>,</span>
<span>  nFeats <span class="op">=</span> <span class="st">"sqrt"</span>,</span>
<span>  featMetric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"F"</span>, <span class="st">"mad"</span>, <span class="st">"sd"</span>, <span class="st">"Sn"</span><span class="op">)</span>,</span>
<span>  DGEmethod <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"limma"</span>, <span class="st">"mast"</span><span class="op">)</span>,</span>
<span>  DGEexpThreshold <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  recalcDataMatrix <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  nBoots <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  useCors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  clustFunc <span class="op">=</span> <span class="st">"cKmeansDownsampleSqrt"</span>,</span>
<span>  clustList <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  linkage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mcquitty"</span>, <span class="st">"ward.D"</span>, <span class="st">"ward.D2"</span>, <span class="st">"single"</span>, <span class="st">"complete"</span>, <span class="st">"average"</span>,</span>
<span>    <span class="st">"centroid"</span><span class="op">)</span>,</span>
<span>  info <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genesets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  qthresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  cthresh <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  ntotal <span class="op">=</span> <span class="fl">20000</span>,</span>
<span>  ScoreGeneSetMethod <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GSVA"</span>, <span class="st">"AUCELL"</span><span class="op">)</span>,</span>
<span>  oneoff <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  stabThresh <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  geneURL <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genesetURL <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>An object of class matrix, dgCMatrix, Seurat, SingleCellExperiment, or ExpressionSet. For matrix and dgCMatrix, include genes and observations (single-cell/bulk profiles) as rows and columns, respectively.</p></dd>


<dt>cohorts</dt>
<dd><p>Character. The column in meta data of 'object' that has cohort IDs. Default NULL if no cohorts in data.</p></dd>


<dt>eMatDS</dt>
<dd><p>Numeric matrix. A matrix with the same number of observations as 'object' containing normalized expression data to be used in analyses downstream of partitioning results.</p></dd>


<dt>colData</dt>
<dd><p>data.frame. Only used if 'object' is a matrix or dgCMatrix. A data frame with named rows and columns containing observation data for each column in 'object'.</p></dd>


<dt>vehicle</dt>
<dd><p>The value in the cohort variable that contains the ID of observation to use as control. Default NULL if no vehicle to be used.</p></dd>


<dt>variables</dt>
<dd><p>Character. Columns in meta data of 'object' to control for in differential analyses.</p></dd>


<dt>seuAssay</dt>
<dd><p>Character. Name of assay in Seurat object containing expression data for running partitioning algorithm. If cohorts based on clustering, this should be the assay used.</p></dd>


<dt>seuAssayDS</dt>
<dd><p>Character. Name of assay in Seurat object containing expression data normalized expression data to be used in analyses downstream of partitioning algorithm.</p></dd>


<dt>sceAssay</dt>
<dd><p>Character. Name of assay in SingleCellExperimen object containing expression data for running partitioning algorithm. If cohorts based on clustering, this should be the assay used.</p></dd>


<dt>sceAssayDS</dt>
<dd><p>Character. Name of assay in SingleCellExperiment object containing expression data normalized expression data to be used in analyses downstream of partitioning algorithm.</p></dd>


<dt>block</dt>
<dd><p>Character. Block parameter in limma for modeling higherarchical data structure, such as multiple observations per individual.</p></dd>


<dt>logCounts</dt>
<dd><p>Logical. Whether or not expression values are log-scale counts or log normalized counts from RNA-seq. Default is TRUE.</p></dd>


<dt>use</dt>
<dd><p>Character. Options are 'Z' to generate test statistics or 'MEAN' to use means from differential analysis for clustering.</p></dd>


<dt>nFeats</dt>
<dd><p>'sqrt' or a numeric value &lt;= number of features to subset for each partition.</p></dd>


<dt>featMetric</dt>
<dd><p>Character. Metric to use to assign gene-level variance/signal score.</p><ul><li><p>F: F-statistic from evaluating differences in means across cohort</p></li>
<li><p>mad: Median absolute deviation</p></li>
<li><p>sd: Standard deviation</p></li>
<li><p>Sn: Robust scale estimator</p></li>
</ul></dd>


<dt>DGEmethod</dt>
<dd><p>Character. Method for running differential gene expression analyses. Use one of either 'limma' (default) or 'mast'.</p></dd>


<dt>DGEexpThreshold</dt>
<dd><p>Numeric. A value between 0 and 1 indicating for filtering lowly expressed genes for partition-specific differential gene expression. Proportion of observations with counts &gt; 0 in at least one subgroup at a specific partition.</p></dd>


<dt>recalcDataMatrix</dt>
<dd><p>Logical. Recalculate dataMatrix for each partion? Default is TRUE.</p></dd>


<dt>nBoots</dt>
<dd><p>nBoots A value of the number of bootstraps to run at each partition. Default is 500.</p></dd>


<dt>useCors</dt>
<dd><p>Numeric. Number of cores to use for parallelizable processes.</p></dd>


<dt>clustFunc</dt>
<dd><p>Character. Wrapper function to be used in recursive partitioning.</p><ul><li><p>cKmeansDownsampleSqrt: Perform constrained K-means clustering after
 subsampling each cohort by the square root of the number of observations</p></li>
<li><p>cKmeansDownsampleSmallest: Perform constrained K-means clustering
 after subsampling each cohort by the size of the smallest cohort</p></li>
<li><p>hclustWrapper: Perform hierarchical clustering</p></li>
</ul></dd>


<dt>clustList</dt>
<dd><p>Optional named list of parameters to use with clustFunc.</p><ul><li><p>cKmeansDownsampleSqrt: 
   </p><ul><li><p>maxIter: The maximum number of iterations to use with lcvqe()</p></li>
</ul></li>
<li><p>cKmeansDownsampleSmallest: 
   </p><ul><li><p>maxIter: The maximum number of iterations to use with lcvqe()</p></li>
</ul></li>
<li><p>hclustWrapper: 
   </p><ul><li><p>aggMethod: One of the hierarchichal methods specified by hclust() function</p></li>
<li><p>distMetric: One of the distance metrics specified by dist() function</p></li>
</ul></li>
</ul></dd>


<dt>linkage</dt>
<dd><p>Character. Linkage criteria for splitting cosine matrix ('method' in hclust). 'average' by default.</p></dd>


<dt>info</dt>
<dd><p>Character. A vector of column names in meta data of 'object' that contain information to be used in cohort annotation of dashboard visualization</p></dd>


<dt>genesets</dt>
<dd><p>Named list. Feature sets to be includes in enrichment-based analyses.</p></dd>


<dt>qthresh</dt>
<dd><p>Numeric. A value between 0 and 1 indicating the FDR cuttoff to define feature sets.</p></dd>


<dt>cthresh</dt>
<dd><p>Numeric. A positive value for the coefficient cuttoff to define feature sets.</p></dd>


<dt>ntotal</dt>
<dd><p>Numeric. A positive value to use as the background feature count. 20000 by default.</p></dd>


<dt>ScoreGeneSetMethod</dt>
<dd><p>Character. Method for gene set scoring. Use one of either 'GSVA' (default) or 'AUCELL'.</p></dd>


<dt>oneoff</dt>
<dd><p>Logical. Allow 1 observation partition groups? Default is TRUE.</p></dd>


<dt>stabThresh</dt>
<dd><p>Numeric. A value between 0 and 1 indicatingThreshold for ending clustering.</p></dd>


<dt>geneURL</dt>
<dd><p>Named list. URLs linking genes to external resources.</p></dd>


<dt>genesetURL</dt>
<dd><p>Named list. URLs linking gene set to external resources.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>An object of class, `K2`.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Reed ER, Monti S (2021).
“Multi-resolution characterization of molecular taxonomies in bulk and single-cell transcriptomics data.”
<em>Nucleic Acids Research</em>.
<a href="https://doi.org/10.1093/nar/gkab552" class="external-link">doi:10.1093/nar/gkab552</a>
, <a href="https://pubmed.ncbi.nlm.nih.gov/34226941/" class="external-link">https://pubmed.ncbi.nlm.nih.gov/34226941/</a>.

Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi W, Smyth GK (2015).
“limma powers differential expression analyses for RNA-sequencing and microarray studies.”
<em>Nucleic Acids Research</em>, <b>43</b>(7), e47--e47.
ISSN 1362-4962, 0305-1048, <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">doi:10.1093/nar/gkv007</a>
, <a href="http://academic.oup.com/nar/article/43/7/e47/2414268/limma-powers-differential-expression-analyses-for" class="external-link">http://academic.oup.com/nar/article/43/7/e47/2414268/limma-powers-differential-expression-analyses-for</a>.

Rousseeuw PJ, Croux C (1993).
“Alternatives to the Median Absolute Deviation.”
<em>Journal of the American Statistical Association</em>, <b>88</b>(424), 1273-1283.
ISSN 0162-1459, <a href="https://doi.org/10.1080/01621459.1993.10476408" class="external-link">doi:10.1080/01621459.1993.10476408</a>
.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Eric Reed.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

